Add advanced setting to enable editing nonce on confirmation screens

rickycodes opened the pull request:
re: #6757

This gives us a new nonce toggle in settings:

![image](https://user-images.githubusercontent.com/675259/64133427-f2e57c80-cda3-11e9-9fcc-8a46ac3d30cb.png)

Which enables adding one to the transaction:

![image](https://user-images.githubusercontent.com/675259/64133458-29bb9280-cda4-11e9-9829-8d89e6b4ce68.png)

rickycodes committed:
Add UseNonce toggle

rickycodes committed:
Get the toggle actually working and dispatching

rickycodes committed:
Display nonce field on confirmation page

rickycodes committed:
Remove console.log

rickycodes committed:
Add placeholder

rickycodes committed:
Set customNonceValue

rickycodes committed:
Add nonce key/value to txParams

rickycodes committed:
remove customNonceValue from component state

rickycodes committed:
Use translation file and existing CSS class

rickycodes committed:
Use existing TextField component

rickycodes committed:
Remove console.log

rickycodes committed:
Fix lint nits

rickycodes commented on ui/app/pages/confirm-transaction-base/confirm-transaction-base.container.js:
```diff
@@ -172,7 +180,13 @@ const mapDispatchToProps = dispatch => {
+    sendTransaction: txData => {
+      const _txData = customNonceValue ? {
+        ...txData,
+        customNonceValue,
```
I tried doing: `nonce: customNonceValue` to see if i could get the nonce into the transaction that way, but it didn't appear to work as expected...

danjm replied:
This is due to how the nonce is set when a transaction is approved. See line 379 of the transactions controller: https://github.com/MetaMask/metamask-extension/pull/7089/files/5e653e49f9376c12a33b77ce5813bb30468c2c33#diff-97c997fde8231c768b7f2393fd3012e1L379

It looks like you were on your way to updating that code based on your comment in that file

rickycodes commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -376,6 +376,7 @@ class TransactionController extends EventEmitter {
       // add nonce to txParams
       // if txMeta has lastGasPrice then it is a retry at same nonce with higher
       // gas price transaction and their for the nonce should not be calculated
+      log.debug('customNonceValue', txMeta.customNonceValue)
```
I'd assume this property exists at this point, but I don't see this log anywhere? not sure what makes the most sense to debug this?

danjm replied:
Were you looking in the background console for the log? To see it you need to:
- go to `chrome://extensions/` or the Brave equivalent
- turn "Developer Mode" on
- under the MetaMask app you are building, inspect the "Background page"
- click the "console" tab

rickycodes replied:
Okay, I got things _sorta_ working... was able to kick off a transaction with a custom nonce of `4`:

![image](https://user-images.githubusercontent.com/675259/64303456-3d0e5f80-cf56-11e9-8775-77bb7dd5785e.png)

however it's just been pending this whole time... I'm not sure what the requirements of `nonce` are... trying strings like 'testing123' (as an example) just seems to break things (transaction doesn't go through and I get a bunch of errors).

danjm replied:
In fact, I think what you have is working. That the tx is in pending the whole time is due to how transactions are handled by the ethereum blockchain. If the network only has record of nonces 1 and 2 for your account, and you custom edit a new transaction to nonce 5, that new one will remain pending until transactions with nonces 3 and 4 are successfully confirmed.

Read this for a helpful explanation: https://www.reddit.com/r/ethereum/comments/6ihw6p/can_someone_please_explain_nonce_to_me/dj6r1lk/

You are right, however, that a nonce cannot just be any arbitrary string. It must be a natural number >= 0. So we should prevent other values from being entered in this input field.

danjm replied:
- [ ] Ensure inputs can only be submitted if nonce is a whole number >= 0, or ensure those are the only input possible in the nonce field

rickycodes reviewed (commented):
This is _almost_ done, just a couple things need to be tied together

rickycodes committed:
Okay this sorta works?

danjm commented:
We should move the toggle to the "Advanced" settings page.

- [ ] move the toggle to the "Advanced" settings page.

rickycodes committed:
Move nonce toggle to advanced tab

rickycodes committed:
Set min to 0

rickycodes committed:
Wrap value in Number()

rickycodes committed:
Add customNonceMap

rickycodes committed:
Update custom nonce translation

rickycodes committed:
Update styles

rickycodes committed:
Reset CustomNonce

rickycodes commented:
I've moved the nonce toggle to the Advanced settings page, I've added some additional styles and things, but I am sure there is still more to do around where to reset the customNonceValue... I've got it resetting when you toggle and when the transaction goes through, but it might need to be sprinkled out a bit more?

rickycodes commented:
Oh also there might need to be more of a sanity check on `customNonceValue` when it's getting set, i don't know if `parseInt` would suffice in this case? `0` is valid, but `001` is invalid (if I understand things correctly)... maybe there's already a thing in the wild for validating nonce? anyway, i don't think setting the input to `type=number` really satisfies the requirements?

rickycodes committed:
Fix lint

rickycodes committed:
Get tests passing

rickycodes committed:
Add customNonceValue to defaults

rickycodes committed:
Fix test

rickycodes committed:
Fix comments

rickycodes committed:
Update tests

rickycodes committed:
Use camel case

fabtho commented:
Thanks for this! I'am looking forward to use it. I got a stuck tx since more then 4 days, metamask not showing speed up or cancel on this one. (Browser freeze, so tx somewhat lost)

danjm committed:
Ensure custom nonce can only be whole number

danjm committed:
Correct font size for custom nonce input

danjm commented:
Updated demo: https://streamable.com/noij3

And screen shot:

![Screenshot from 2019-09-19 09-49-30](https://user-images.githubusercontent.com/7499938/65243586-296d0800-dac3-11e9-8b89-9029ea81f9b3.png)

cjeria commented:
The nonce input field is very wide. Probably could reduce the width by 1/2 it's current size.

![image](https://user-images.githubusercontent.com/597770/65289129-bb4a3480-db06-11e9-91ad-1b69ba120843.png)

I'd update the text in settings to something like this instead.
![image](https://user-images.githubusercontent.com/597770/65289132-bdac8e80-db06-11e9-9610-e7e5c4cc0dd4.png)
cc @danjm

danjm committed:
UX improvements for custom nonce feature

danjm committed:
Fix advanced-tab-component tests for custom nonce changes

danjm commented:
@cjeria Made changes according to your suggestions:

![Screenshot from 2019-09-21 13-53-51](https://user-images.githubusercontent.com/7499938/65376245-472f9e00-dc78-11e9-8e69-6ef09f275ecc.png)
-
![Screenshot from 2019-09-21 13-50-59](https://user-images.githubusercontent.com/7499938/65376246-47c83480-dc78-11e9-8d71-ec3e8567a401.png)

Gudahtt commented on app/_locales/en/messages.json:
```diff
@@ -195,6 +195,21 @@
+  "nonceFieldDescription": {
+    "message": "Turn this on to change the nonce (transaction number) on confirmation screens. This is an advanced feature, use cautiously."
+  },
+  "borrowDharma": {
```
It looks like this message was brought back accidentally in the merge conflict resolution - I deleted it recently.

Gudahtt commented on ui/app/pages/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -251,6 +258,26 @@ export default class ConfirmTransactionBase extends Component {
+                  min="0"
+                  placeholder={ this.context.t('nonceFieldPlaceholder') }
+                  onChange={({ target: { value }}) => {
+                    updateCustomNonce(value < 0 ? 0 : Math.floor(value))
```
The custom nonce is a string everywhere else; it might be better to cast it to a string here as well. Better to avoid it having multiple types if possible, just to keep things simpler.

Gudahtt commented on ui/app/store/actions.js:
```diff
@@ -2591,6 +2603,24 @@ function setUseBlockie (val) {
+    log.debug(`background.setUseNonceField`)
+    background.setUseNonceField(val, (err) => {
+      dispatch(actions.hideLoadingIndication())
+      !val && dispatch(actions.updateCustomNonce(''))
```
I'm guessing this line was used to coerce `undefined`/`null` back to a string? Casting to a string in the handler would mean this line isn't needed anymore I suspect, as it'd always be a string.

Gudahtt commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -377,8 +377,10 @@ class TransactionController extends EventEmitter {
       // if txMeta has lastGasPrice then it is a retry at same nonce with higher
       // gas price transaction and their for the nonce should not be calculated
+      const { customNonceValue = null } = txMeta
       const nonce = txMeta.lastGasPrice ? txMeta.txParams.nonce : nonceLock.nextNonce
```
It looks like we're using `nonceLock.nextNonce` here and not using it, in the case where `txMeta.lastGasPrice` is falsy and `customNonceValue` is set

Gudahtt replied:
Actually I think the damage is done up on line 376, not on this line.

Gudahtt replied:
The `nonceLock` could still be referenced here in the custom nonce case (if `lastGasPrice` is truthy).

cjeria commented:
There's a lot of emphasis on the component "nonce field" in the title. Titles should help answer user questions like what the feature will enable them to do. In this case, the "nonce field" lets users "**customize transaction nonces**".  More broadly this feature will let users "**Edit transaction nonces on confirm screens**". 

Also, remove the period at the end of the title. WDYT?

danjm committed:
Update title of nonce toggle in settings

danjm committed:
Remove unused locale message

danjm committed:
Cast custom nonce to string in confirm-transaction-base.component

danjm committed:
Handle string conversion and invalid values for custom nonces in handler

danjm committed:
Don't call getNonceLock in tx controller if there is a custom nonce

danjm committed:
Set nonce details for cases where nonce is customized

danjm committed:
Fix incorrectly use value for deciding whether to getnoncelock in approveTransaction

danjm committed:
Default nonceLock to empty object in approveTransaction

frankiebee commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -365,28 +365,36 @@ class TransactionController extends EventEmitter {
       const nonce = txMeta.lastGasPrice ? txMeta.txParams.nonce : nonceLock.nextNonce
-      txMeta.txParams.nonce = ethUtil.addHexPrefix(nonce.toString(16))
+      const customOrNonce = customNonceValue || nonce
+      txMeta.txParams.nonce = ethUtil.addHexPrefix(customOrNonce.toString(16))
```
is there some where we check that the custom nonce is not a NaN?

Gudahtt replied:
I believe it's impossible for it to be NaN. It comes from an `<input type="number">` field, and it's set to an empty string if it's not a positive number.

frankiebee commented:
QUESTION: should we address the showing of a warning if the custom nonce is distance one or greater from our suggested nonce

danjm committed:
Reapply use on nonceLock in cases where customNonceValue in approveTransaction.

danjm committed:
Show warning message if custom nonce is higher than MetaMask's next nonce

danjm committed:
Fix e2e test failure caused by custom nonce and 3box toggle conflict

Gudahtt commented on ui/app/pages/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -251,6 +263,37 @@ export default class ConfirmTransactionBase extends Component {
+                        this.setState({ submitWarning: '' })
+                      }
+                      updateCustomNonce(String(newCustomNonce))
+                      getNextNonce()
```
If this results in the nextNonce changing to one that would invalidate the custom nonce, it won't result in a warning. I think this case could be caught by re-validating the custom nonce field in the lifecycle function when the `nextNonce` prop changes

danjm committed:
Update nonce warning message to include the suggested nonce

danjm committed:
Handle nextNonce comparison and update logic in lifecycle

danjm committed:
Default nonce field to suggested nonce

danjm committed:
Clear custom nonce on reject or confirm

danjm committed:
Fix bug where nonces are not shown in tx list on self sent transactions

danjm committed:
Ensure custom nonce is reset after tx is created in background

danjm committed:
Convert customNonceValue to number in approve tranasction controller

danjm committed:
Lint fix

danjm committed:
Call getNextNonce after updating custom nonce

frankiebee reviewed (approved):
Thanks looks great

tbenr commented:
@danjm this is not implemented for ERC20 "approve" transactions, am I right?
